src/ethereum-canister/tests/test_canister/mod.rs

```diff
@@ -0,0 +1,125 @@
+use std::fs::File;
+
+use candid::{utils::ArgumentEncoder, IDLArgs};
+use eyre::{ensure, Result, WrapErr};
+use serde_json::{json, Value};
+use temp_dir::TempDir;
+
+static DFX_JSON: &str = include_str!("../../../../dfx.json");
+
+const DEFAULT_CONSENSUS_RPC: &str = "https://www.lightclientdata.org";
+const DEFAULT_EXECUTION_RPC: &str = "https://ethereum.publicnode.com";
+
+// NOTE: we can't just decode from inside the `call` because decode_args
+// returns ArgumentDecoder<'a> where 'a lifetime would be tied to the output we're
+// parsing, which would be dropped on return.
+// The reason for this is that ArgumentDecoder allows decoding by ref eg. to (&str,)
+macro_rules! call_decode {
+    ($canister:expr, $method:expr, $args:expr) => {{
+        $canister.call($method, $args).and_then(|output| {
+            ::candid::utils::decode_args(&output).map_err(|err| ::eyre::eyre!(err))
+        })
+    }};
+}
+
+pub(crate) use call_decode;
+
+#[derive(Debug)]
+pub struct TestCanister {
+    name: String,
+    temp_dir: TempDir,
+}
+
+impl TestCanister {
+    pub fn deploy() -> Self {
+        let temp_dir = TempDir::new().unwrap();
+
+        // setup the tempdir
+        make_symlink(&temp_dir, "src");
+        make_symlink(&temp_dir, "target");
+        make_symlink(&temp_dir, "Cargo.toml");
+        make_symlink(&temp_dir, "Cargo.lock");
+
+        let name = format!("ethereum_canister_{}", rand::random::<u64>());
+
+        // rename the canister in dfx.json
+        let mut dfx_config: Value = serde_json::from_str(DFX_JSON).unwrap();
+        let ethereum_canister = dfx_config["canisters"]["ethereum_canister"].take();
+        dfx_config["canisters"] = json!({ name.as_str(): ethereum_canister });
+
+        // create dfx.json
+        let dfx_json = File::create(temp_dir.child("dfx.json")).unwrap();
+        serde_json::to_writer_pretty(dfx_json, &dfx_config).unwrap();
+
+        // deploy
+        let canister = TestCanister { temp_dir, name };
+        canister.run_dfx(&["deploy"]).unwrap();
+
+        canister
+    }
+
+    pub fn deploy_and_setup() -> Self {
+        let canister = Self::deploy();
+        let request = interface::SetupRequest {
+            consensus_rpc_url: DEFAULT_CONSENSUS_RPC.to_owned(),
+            execution_rpc_url: DEFAULT_EXECUTION_RPC.to_owned(),
+        };
+        canister.call("setup", (request,)).unwrap();
+        canister
+    }
+
+    pub fn call<Args: ArgumentEncoder>(&self, method: &str, args: Args) -> Result<Vec<u8>> {
+        let args = candid::utils::encode_args(args).wrap_err("encoding args")?;
```

Comment:
```suggestion
    pub fn call<C: CandidType>(&self, method: &str, arg: C) -> Result<Vec<u8>> {
        let args = candid::utils::encode_one(args).wrap_err("encoding args")?;
```
If we make such a change to this method, then we can pass single arguments without wrapping them in `(x,)`, like
```rs
let res: (String,) = call_decode!(canister, "foo", request).unwrap();
```
but we will no longer have ability to call canister queries that have more than one argument, because eg.
`(15, 12)` will be encoded as `(record { 15 : nat, 12 : nat })` which is a single argument.

Reply:
I came up with [24db9c7](https://github.com/eigerco/ethereum-canister/pull/29/commits/24db9c7264af94fa9fe2c83d30bbcd7fb8225051) to address that